kernel/rtlil.h

```diff
@@ -657,38 +667,111 @@ namespace RTLIL
 
 struct RTLIL::Const
 {
-	int flags;
-	std::vector<RTLIL::State> bits;
+	short flags;
+	private:
+	typedef std::vector<RTLIL::State> bitvectype;
+	enum class backing_tag: bool { bits, string };
+	// Do not access the union or tag even in Const methods unless necessary
+	mutable backing_tag tag;
+	union {
+		mutable bitvectype bits_;
+		mutable std::string str_;
+	};
+
+	// Use these utilities instead
+	bool is_bits() const { return tag == backing_tag::bits; }
+	bool is_str() const { return tag == backing_tag::string; }
 
-	Const() : flags(RTLIL::CONST_FLAG_NONE) {}
+	bitvectype* get_if_bits() const { return is_bits() ? &bits_ : NULL; }
+	std::string* get_if_str() const { return is_str() ? &str_ : NULL; }
+
+	void check(bool condition) const {
+		// TODO debug assert or throw
+		// log_assert(condition && "malformed Const union");
+	}
```

Comment:
I am not convinced we want this helper. This is worse than `log_assert` at the call site unless we start inspecting backtrace here

Reply:
How about this hack:

```diff
+#define check(condition) log_assert(condition && "malformed Const union")
 struct RTLIL::Const
 {
        short flags;
@@ -685,11 +686,6 @@ struct RTLIL::Const
        bitvectype* get_if_bits() const { return is_bits() ? &bits_ : NULL; }
        std::string* get_if_str() const { return is_str() ? &str_ : NULL; }

-       void check(bool condition) const {
-               // TODO debug assert or throw
-               // log_assert(condition && "malformed Const union");
-       }
-
        bitvectype& get_bits() const { check(is_bits()); return *get_if_bits(); }
        std::string& get_str() const { check(is_str()); return *get_if_str(); }
        public:
@@ -816,6 +812,7 @@ struct RTLIL::Const
                return h;
        }
 };
+#undef check
```

Thoughts? I previously verified with benchmarks that keeping the assert is free. Those asserts will make everything so verbose

Reply:
I'm fine with it